# **Problem 3**

In [ ]:
from google.colab import files
import numpy as np
import pandas as pd
import statsmodels.api as sm
from statsmodels.formula.api import ols
from sklearn.linear_model import LinearRegression
import scipy.stats as st

/usr/local/lib/python3.7/dist-packages/statsmodels/tools/_testing.py:19: FutureWarning: pandas.util.testing is deprecated. Use the functions in the public API at pandas.testing instead.
  import pandas.util.testing as tm


In [ ]:
import matplotlib.pyplot as plt
import seaborn as sns
import io

In [ ]:
uploaded=files.upload()

Saving Carseats.csv to Carseats.csv


In [ ]:
for fn in uploaded.keys():
 print('file"{name}" length{length}'.format(name=fn,length=len(uploaded[fn])))
car=pd.read_csv(io.StringIO(uploaded['Carseats.csv'].decode('utf-8')))

file"Carseats.csv" length17023


In [ ]:
car.head()

Sales  CompPrice  Income  Advertising  ...  Age  Education Urban   US
0   9.50        138      73           11  ...   42         17   Yes  Yes
1  11.22        111      48           16  ...   65         10   Yes  Yes
2  10.06        113      35           10  ...   59         12   Yes  Yes
3   7.40        117     100            4  ...   55         14   Yes  Yes
4   4.15        141      64            3  ...   38         13   Yes   No

[5 rows x 11 columns]

**a. Fit a multiple regression model to predict Sales using Price, Advertising Urban, and US.Write out the model in equation form, being careful to handle the qualitative variables properly.**

In [ ]:
X = car[['Price' ,'Urban', 'US']]
y = car['Sales']

In [ ]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2)

In [ ]:
reg = LinearRegression(fit_intercept=True)
reg = ols(formula = 'Sales ~ Price + C(Urban) + C(US)', data = car).fit() 

In [ ]:
reg.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                  Sales   R-squared:                       0.239
Model:                            OLS   Adj. R-squared:                  0.234
Method:                 Least Squares   F-statistic:                     41.52
Date:                Sat, 23 Oct 2021   Prob (F-statistic):           2.39e-23
Time:                        21:05:56   Log-Likelihood:                -927.66
No. Observations:                 400   AIC:                             1863.
Df Residuals:                     396   BIC:                             1879.
Df Model:                           3                                         
Covariance Type:            nonrobust                                         
===================================================================================
                      coef    std err          t      P>|t|      [0.025      0.975]
-----------------------------------------------------------------------------------
Intercept          13.0435      0.651     20.036      0.000      11.764      14.323
C(Urban)[T.Yes]    -0.0219      0.272     -0.081      0.936      -0.556       0.512
C(US)[T.Yes]        1.2006      0.259      4.635      0.000       0.691       1.710
Price              -0.0545      0.005    -10.389      0.000      -0.065      -0.044
==============================================================================
Omnibus:                        0.676   Durbin-Watson:                   1.912
Prob(Omnibus):                  0.713   Jarque-Bera (JB):                0.758
Skew:                           0.093   Prob(JB):                        0.684
Kurtosis:                       2.897   Cond. No.                         628.
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

a. Equation Form.
 Estimated carseat sale=13.043−0.0219X1+1.2006X2−0.0545X3 

    

**b.  Provide an interpretation of each coefficient in the model. Be careful–some of the vari-
ables in the model are qualitative!**

The coefficient of -0.0545 for Price means increasing the price of a car seat by $1 results in a decrease of sales by approximately 54.5 units for 1000 units 

The coefficient of -0.0219 for UrbanYes means that, carseat have approximately 21.9 fewer carseat sales in urban area compared to non-urban areas. 

The coefficient of 1.2006 for USYes means that stores in the United States have 1200.6 more carseat sales on average than stores outside the United States

 **C. Using the model from (a), predict sales in the new store and calculate 68% and 95%
confidence intervals.**


*   Intercept: (11.764, 14.271)
*   Urban: (-0.556,0.512)
*   US: (0.692, 1.708)
*   Price: (-0.065, -0.044)

In [ ]:
y_pred = reg.predict(X_test)

In [ ]:
X_test.head()

Price Urban   US
293     89   Yes   No
112    125   Yes  Yes
206    160   Yes  Yes
223    125   Yes  Yes
16     110   Yes   No

In [ ]:
y_pred

293    8.174715
112    7.414769
206    5.508710
223    7.414769
16     7.031079
         ...   
53     7.741522
92     6.214197
359    7.142475
202    7.164391
96     7.055474
Length: 80, dtype: float64

In [ ]:
y_test

293    11.28
112     6.67
206     4.97
223     3.45
16      7.58
       ...  
53      6.92
92      4.53
359     3.13
202     4.10
96      9.48
Name: Sales, Length: 80, dtype: float64

In [ ]:
y = car['Sales']

In [ ]:
CI1= st.t.interval(0.95, df=len(y),scale=st.sem(y))
CI2= st.t.interval(0.68, df=len(y),scale=st.sem(y))
print('95% CI of Sales:',CI1) 
print('68% CI of sales:',CI2) 

95% CI of Sales: (-0.2775981316574618, 0.2775981316574618)
68% CI of sales: (-0.14059794945210255, 0.14059794945210255)


In [ ]:
print('68% Confidence Interval')
print('Intercept :({low},{high})'.format(low=13.0435-(1*0.651),high=13.0435+(1*0.651)))
print('Urban :({low},{high})'.format(low=	-0.0219-(1*0.272),high=	-0.0219+(1*0.272)))    
print('US :({low},{high})'.format(low=1.2006-(1*0.259),high=1.2006+(1*0.259)))  
print('Price :({low},{high})'.format(low=-0.0545-(1*0.005),high=-0.0545+(1*0.005)))  

68% Confidence Interval
Intercept :(12.3925,13.6945)
Urban :(-0.2939,0.25010000000000004)
US :(0.9415999999999999,1.4596)
Price :(-0.0595,-0.0495)


**d.  Using the model from (a), what is the probability that sales will be greater than 12000units in the new store?**

In [ ]:
mean=car['Sales'].mean()
std=car['Sales'].std()
zval=(1.2000-mean)/std
print('Probability that sales will be greater than 12000 units:',st.norm.cdf(zval))

Probability that sales will be greater than 12000 units: 0.012890792467909518


**e. Using the model from (a), what is the probability that sales will be between 6000 and 10000 units in the new store?**

In [ ]:
mean=car['Sales'].mean()
std=car['Sales'].std()
zval1=(0.6000-mean)/std
zval2=(1.0000-mean)/std
z1=st.norm.cdf(zval2)
z2=st.norm.cdf(zval1)
print('Probability that sales will be between 6000 and 10000 units:',z1-z2)

Probability that sales will be between 6000 and 10000 units: 0.003411235225023271


**f. For which of the predictors can you reject the null hypothesis H0:βj= 0?**

The p-values for the intercept, Price, and USYes are all zero < 0.005, which provides strong evidence to reject the null hypothesis  H0:βj=0.
The p-value for UrbanYes, however, is 0.936 which is  > 0.005, so there is no evidence to reject the null hypothesis.

**g. On the basis of your response to the previous question,  fit a smaller model that onlyuses the predictors for which there is evidence of association with the outcome.  Using this model,predict sales in the new store and calculate 68% and 95% confidence intervals.**

In [ ]:
XNew = car[['Price' , 'US']]
yNew = car['Sales']

In [ ]:
from sklearn.model_selection import train_test_split
X_trainNew, X_testNew, y_trainNew, y_testNew = train_test_split(XNew, yNew, test_size=0.2)

In [ ]:
regNew = LinearRegression(fit_intercept=True)
regNew = ols(formula = 'Sales ~ Price + C(US)', data = car).fit() 
regNew.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                  Sales   R-squared:                       0.239
Model:                            OLS   Adj. R-squared:                  0.235
Method:                 Least Squares   F-statistic:                     62.43
Date:                Sat, 23 Oct 2021   Prob (F-statistic):           2.66e-24
Time:                        19:49:53   Log-Likelihood:                -927.66
No. Observations:                 400   AIC:                             1861.
Df Residuals:                     397   BIC:                             1873.
Df Model:                           2                                         
Covariance Type:            nonrobust                                         
================================================================================
                   coef    std err          t      P>|t|      [0.025      0.975]
--------------------------------------------------------------------------------
Intercept       13.0308      0.631     20.652      0.000      11.790      14.271
C(US)[T.Yes]     1.1996      0.258      4.641      0.000       0.692       1.708
Price           -0.0545      0.005    -10.416      0.000      -0.065      -0.044
==============================================================================
Omnibus:                        0.666   Durbin-Watson:                   1.912
Prob(Omnibus):                  0.717   Jarque-Bera (JB):                0.749
Skew:                           0.092   Prob(JB):                        0.688
Kurtosis:                       2.895   Cond. No.                         607.
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

In [ ]:
y_predNew = regNew.predict(X_testNew)
X_testNew.head()

Price   US
210    123  Yes
18      68  Yes
106    139   No
343    121  Yes
329     89  Yes

In [ ]:
y_predNew

210     7.529687
18     10.525957
106     5.458402
343     7.638642
329     9.381926
         ...    
298     5.948701
99      8.401329
87      7.366254
240     6.438999
184     8.946105
Length: 80, dtype: float64

In [ ]:
y_testNew

210     4.36
18     13.91
106     0.16
343     5.99
329    11.27
       ...  
298    10.98
99      4.88
87     11.70
240    10.31
184     9.95
Name: Sales, Length: 80, dtype: float64

**95% confidence intervals:**


*   Intercept: (11.790, 14.271)
*   US: (0.692, 1.708)
*   Price: (-0.065, -0.044)

In [ ]:
CI1= st.t.interval(0.95, df=len(yNew),scale=st.sem(yNew))
CI2= st.t.interval(0.68, df=len(yNew),scale=st.sem(yNew))
print('95% CI of Sales:',CI1) 
print('68% CI of sales:',CI2) 

95% CI of Sales: (-0.2775981316574618, 0.2775981316574618)
68% CI of sales: (-0.14059794945210255, 0.14059794945210255)


In [ ]:
print('68% Confidence Interval')
print('Intercept :({low},{high})'.format(low=13.0435-(1*0.651),high=13.0435+(1*0.651)))
print('US :({low},{high})'.format(low=1.2006-(1*0.259),high=1.2006+(1*0.259)))  
print('Price :({low},{high})'.format(low=-0.0545-(1*0.005),high=-0.0545+(1*0.005)))


68% Confidence Interval
Intercept :(12.3925,13.6945)
US :(0.9415999999999999,1.4596)
Price :(-0.0595,-0.0495)


**h. How well do the models in (a) and (g) fit the data?**

The models in a and g both fit the data well, with identical  R-Squared  values of 0.2393 and intercept as 13.04.